src/Instrumentation/ExtRdKafka/tests/Integration/ExtRdKafkaInstrumentationTest.php

```diff
@@ -0,0 +1,202 @@
+
+    public function tearDown(): void
+    {
+        $this->scope->detach();
+    }
+
+    public function test_consume_creates_new_span(): void
+    {
+        // Given
+        $this->produceMessage('test', produceWithoutHeaders: true);
+        $this->assertCount(0, $this->storage);
+
+        // When
+        $this->consumeMessage();
+
+        // Then
+        $this->assertCount(1, $this->storage);
+        /** @var ImmutableSpan $span */
+        $span = $this->storage->offsetGet(0);
+        $this->assertEquals('test process', $span->getName());
+    }
+
+    public function test_context_propagated_on_consumption(): void
+    {
+        // Give, produceWithoutHeaders: true
+        $tracerProvider = Globals::tracerProvider();
+        $tracer = $tracerProvider->getTracer('test');
+        $span = $tracer->spanBuilder('test_span')->startSpan();
+        $scope = $span->activate();
+
+        $this->produceMessage('test');
+
+        $scope->detach();
+        $span->end();
+        $expectedTraceId = $span->getContext()->getTraceId();
+
+        // One span that I had to create in the test, so I can ensure that the message is produced with a TraceID
+        // that we know about in the test (for assertions). Another for the actual production of the message
+        $this->assertCount(2, $this->storage);
+
+        // When
+        // This should create a third span, for consumption, which should have the same TraceID as the first span
+        $this->consumeMessage();
+
+        // Then
+        $this->assertCount(3, $this->storage);
+        /** @var ImmutableSpan $span */
+        $span = $this->storage->offsetGet(2);
+        $this->assertEquals('test process', $span->getName());
+        $this->assertEquals($expectedTraceId, $span->getContext()->getTraceId());
+    }
+
+    public function test_context_set_in_kafka_headers_on_message_production(): void
+    {
+        // Given
+        $this->assertCount(0, $this->storage);
+        $tracerProvider = Globals::tracerProvider();
+        $tracer = $tracerProvider->getTracer('test');
+        $span = $tracer->spanBuilder('test_span')->startSpan();
+        $scope = $span->activate();
+
+        $this->produceMessage('test');
+
+        $scope->detach();
+        $span->end();
+
+        // When
+        $message = $this->consumeMessage();
+        $this->assertIsArray($message->headers);
+        $this->assertArrayHasKey('traceparent', $message->headers);
+
+        // The traceparent header is separated into 4 sections with dashes. I want to get the second segment of that
+        // header, which is the TraceID
+        $traceId = explode('-', $message->headers['traceparent'])[1];
+        $this->assertEquals($span->getContext()->getTraceId(), $traceId);
+    }
+
+    public function test_produce_creates_new_span()
+    {
+        $this->assertCount(0, $this->storage);
+        // Overriding topic as all other tests consume their own message. If we don't do this, the message will be
```

Comment:
this sounds like a post-test cleanup/teardown issue - could a `tearDown` method empty all queues?

Reply:
Yep, it is exactly that. Had a bit of a look around yesterday, and unfortunately the `ext-rdkafka` extension [doesn't have support for the AdminClient](https://github.com/arnaud-lb/php-rdkafka/issues/215) (the thing used to manage topics) yet.

So from what I can see, there's no way that we can manually remove the topic (or messages, as deletion is handled by the Kafka broker) after a given test run- atleast via PHP. Open to any other ideas, but without having that element of control via the broker, I'm struggling to come up with something better

Reply:
can you consume and discard any remaining messages from all queues as part of teardown?

Reply:
Yeah that will also work, leaves it in a slightly better state than what we've currently got IMO.... Unfortunate that we can't make these tests properly atomic, but I think that's the best we'll get.

Will push up a change asap